File: jetbrains/src/main/kotlin/identitybraces/editor/scenes/SwapPlacesScene.kt
package identitybraces.editor.scenes

import identitybraces.core.TraitIds
import identitybraces.editor.scenes.Keyframes.Key
import java.awt.AlphaComposite
import java.awt.Font
import java.awt.Graphics2D
import java.awt.Rectangle
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Two braces on the same line visit each other's column, and come back.
 *
 * The braces themselves stay put; each sends a likeness — a copy of its own glyph, in its
 * own colour — which arcs across to the other's column and back while the originals hold
 * their ground. Lifting the copies clear of the line as they cross is what stops them
 * reading as a rendering fault: two glyphs sliding *through* the text between them would
 * look like a redraw bug; two glyphs hopping over it looks deliberate.
 */
class SwapPlacesScene : Scene {
    override val traitId: String = TraitIds.SWAP_PLACES
    override val subjectTraitId: String? = null
    override val durationMs: Long = 2400

    /**
     * Finds two candidates sharing a line at a sensible distance. The director resolves
     * candidates as a consecutive run through the brace map, which is what makes this likely
     * to succeed: consecutive braces are usually neighbours on a line.
     */
    override fun tryCast(actors: List<SceneActor>, subjects: List<SceneActor>): List<SceneActor>? {
        for (i in actors.indices) {
            for (j in i + 1 until actors.size) {
                if (actors[i].lineStart != actors[j].lineStart) {
                    continue
                }

                val separation = abs(actors[i].column - actors[j].column)
                if (separation < MIN_SEPARATION || separation > MAX_SEPARATION) {
                    continue
                }

                return listOf(actors[i], actors[j])
            }
        }

        return null
    }

    override fun bounds(cast: List<SceneActor>): Rectangle {
        val left = min(cast[0].cellLeft, cast[1].cellLeft) - cast[0].cellWidth
        val right = max(cast[0].cellLeft, cast[1].cellLeft) + cast[0].cellWidth * 2
        val top = cast[0].textTop - cast[0].textHeight * 1.2
        return Rectangle(left.toInt(), top.toInt(), (right - left).toInt(), (cast[0].textHeight * 3.4).toInt())
    }

    override fun paint(g: Graphics2D, cast: List<SceneActor>, progress: Double, glyphFont: Font) {
        // Arcing in opposite directions, so the two never occupy the same air. Sent the same
        // way they would collide in the middle, which reads as one glyph rather than as two
        // passing.
        send(g, cast[0], cast[1], -1.0, progress, glyphFont)
        send(g, cast[1], cast[0], 1.0, progress, glyphFont)
    }

    /** Draws [from]'s likeness on its way to [to] and back. */
    private fun send(g: Graphics2D, from: SceneActor, to: SceneActor, lift: Double, t: Double, font: Font) {
        val travel = to.cellLeft - from.cellLeft
        val height = from.textHeight * 0.95 * lift

        val dx = Keyframes.at(
            t,
            Key(0.0, 0.0),
            Key(0.40, travel, Keyframes::sineInOut),
            Key(0.60, travel),
            Key(0.95, 0.0, Keyframes::sineInOut),
        )

        val dy = Keyframes.at(
            t,
            Key(0.0, 0.0),
            Key(0.20, height, Keyframes::sineOut),
            Key(0.40, 0.0, Keyframes::sineIn),
            Key(0.77, height, Keyframes::sineOut),
            Key(0.95, 0.0, Keyframes::sineIn),
        )

        // Faded at both ends, so a likeness never sits on top of the original it copied —
        // which would show as one brace suddenly rendering twice as boldly.
        val opacity = Keyframes.at(t, Key(0.0, 0.0), Key(0.12, 1.0), Key(0.85, 1.0), Key(0.96, 0.0))
        if (opacity <= 0.01) {
            return
        }

        val g2 = g.create() as Graphics2D
        try {
            g2.composite = AlphaComposite.SrcOver.derive(opacity.toFloat())
            g2.font = font
            g2.color = from.color
            g2.drawString(from.character.toString(), (from.cellLeft + dx).toFloat(), (from.baselineY - dy).toFloat())
        } finally {
            g2.dispose()
        }
    }

    companion object {
        /**
         * Furthest apart, in columns, that two braces will bother to trade. A swap across
         * forty columns is two things moving independently at opposite ends of the screen,
         * not a pair trading.
         */
        private const val MAX_SEPARATION = 24

        /** Nearest, in columns. Adjacent braces just look like a wobble. */
        private const val MIN_SEPARATION = 2
    }
}

File: src/IdentityBraces/Adornments/Scenes/SwapPlacesScene.cs
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;
using IdentityBraces.Core;

namespace IdentityBraces.Adornments.Scenes
{
    /// <summary>
    /// Two braces on the same line visit each other's column, and come back.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The one scene where the ownership rule bites. The obvious staging is for the two braces
    /// themselves to slide past each other, and a scene may not do that: their visuals belong
    /// to <see cref="AdornmentManager"/>, which rewrites their coordinates on every layout
    /// pass. So each brace sends a likeness instead — a copy of its own glyph, in its own
    /// colour, which arcs across to the other's column and back while the originals hold
    /// their ground.
    /// </para>
    /// <para>
    /// It reads because the copies are the same glyph in the same colour: what you see is two
    /// braces briefly occupying each other's spot. The originals staying put is visible if you
    /// look for it, and the alternative was a mechanism where two components could each move
    /// the same element.
    /// </para>
    /// <para>
    /// Lifting the copies clear of the line as they cross is what stops them reading as a
    /// rendering fault. Two glyphs sliding <em>through</em> the text between them would look
    /// like a redraw bug; two glyphs hopping over it looks deliberate.
    /// </para>
    /// </remarks>
    internal sealed class SwapPlacesScene : IScene
    {
        /// <summary>Furthest apart, in columns, that two braces will bother to trade.</summary>
        /// <remarks>
        /// A swap across forty columns is two things moving independently at opposite ends of
        /// the screen, not a pair trading. Close enough to see both at once is the point.
        /// </remarks>
        private const int MaxSeparation = 24;

        /// <summary>Nearest, in columns. Adjacent braces just look like a wobble.</summary>
        private const int MinSeparation = 2;

        public string TraitId
        {
            get { return TraitIds.SwapPlaces; }
        }

        public string SubjectTraitId
        {
            get { return null; }
        }

        public TimeSpan Duration
        {
            get { return TimeSpan.FromSeconds(2.4); }
        }

        /// <summary>
        /// Finds two candidates sharing a line at a sensible distance.
        /// </summary>
        /// <remarks>
        /// The director resolves candidates as a consecutive run through the brace map rather
        /// than at random, which is what makes this likely to succeed: consecutive braces are
        /// usually neighbours on a line. Without that, two independently chosen braces on a
        /// screen would almost never share one.
        /// </remarks>
        public bool TryCast(IList<SceneActor> actors, IList<SceneActor> subjects, List<SceneActor> cast)
        {
            for (int i = 0; i < actors.Count; i++)
            {
                for (int j = i + 1; j < actors.Count; j++)
                {
                    if (actors[i].LineStart != actors[j].LineStart)
                    {
                        continue;
                    }

                    int separation = Math.Abs(actors[i].Column - actors[j].Column);
                    if (separation < MinSeparation || separation > MaxSeparation)
                    {
                        continue;
                    }

                    cast.Add(actors[i]);
                    cast.Add(actors[j]);
                    return true;
                }
            }

            return false;
        }

        public void Play(SceneStage stage, IList<SceneActor> cast)
        {
            SceneActor left = cast[0];
            SceneActor right = cast[1];

            // Arcing in opposite directions, so the two never occupy the same air. Sent the
            // same way they would collide in the middle, which reads as one glyph rather than
            // as two passing.
            Send(stage, left, right, -1);
            Send(stage, right, left, 1);
        }

        /// <summary>Sends <paramref name="from"/>'s likeness to <paramref name="to"/> and back.</summary>
        private void Send(SceneStage stage, SceneActor from, SceneActor to, double lift)
        {
            char character = from.Column <= to.Column ? '{' : '}';

            TextBlock copy = stage.GlyphOf(character, from.Color);
            stage.Add(copy, from.CellLeft, from.TextTop);

            var move = new TranslateTransform();
            copy.RenderTransform = move;

            double travel = to.CellLeft - from.CellLeft;
            double height = from.TextHeight * 0.95 * lift;
            var period = new Duration(Duration);

            var across = new DoubleAnimationUsingKeyFrames { Duration = period };
            across.KeyFrames.Add(new LinearDoubleKeyFrame(0, KeyTime.FromPercent(0.00)));
            across.KeyFrames.Add(new EasingDoubleKeyFrame(travel, KeyTime.FromPercent(0.40))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseInOut },
            });
            across.KeyFrames.Add(new LinearDoubleKeyFrame(travel, KeyTime.FromPercent(0.60)));
            across.KeyFrames.Add(new EasingDoubleKeyFrame(0, KeyTime.FromPercent(0.95))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseInOut },
            });

            var over = new DoubleAnimationUsingKeyFrames { Duration = period };
            over.KeyFrames.Add(new LinearDoubleKeyFrame(0, KeyTime.FromPercent(0.00)));
            over.KeyFrames.Add(new EasingDoubleKeyFrame(height, KeyTime.FromPercent(0.20))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseOut },
            });
            over.KeyFrames.Add(new EasingDoubleKeyFrame(0, KeyTime.FromPercent(0.40))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseIn },
            });
            over.KeyFrames.Add(new EasingDoubleKeyFrame(height, KeyTime.FromPercent(0.77))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseOut },
            });
            over.KeyFrames.Add(new EasingDoubleKeyFrame(0, KeyTime.FromPercent(0.95))
            {
                EasingFunction = new SineEase { EasingMode = EasingMode.EaseIn },
            });

            // Faded at both ends, so a likeness never sits on top of the original it copied —
            // which would show as one brace suddenly rendering twice as boldly.
            var visible = new DoubleAnimationUsingKeyFrames { Duration = period };
            visible.KeyFrames.Add(new LinearDoubleKeyFrame(0.0, KeyTime.FromPercent(0.00)));
            visible.KeyFrames.Add(new LinearDoubleKeyFrame(1.0, KeyTime.FromPercent(0.12)));
            visible.KeyFrames.Add(new LinearDoubleKeyFrame(1.0, KeyTime.FromPercent(0.85)));
            visible.KeyFrames.Add(new LinearDoubleKeyFrame(0.0, KeyTime.FromPercent(0.96)));

            stage.Animate(move, TranslateTransform.XProperty, across);
            stage.Animate(move, TranslateTransform.YProperty, over);
            stage.Animate(copy, UIElement.OpacityProperty, visible);
        }
    }
}
